Implement _most_ of the EA interface for DTA/TDA

jbrockmendel opened the pull request:
@TomAugspurger as promised, this implements a handful of tests, and a few more can be ported from #23415, but the EA-specific tests haven't been started.

The big missing piece is DatetimeArray._from_sequence, which I'm getting started on now in a new branch.

Closes #23586

jbrockmendel committed:
implement most of the rest of EA interface

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into dlike_ea4

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into dlike_ea4

jbrockmendel committed:
implement some tests for take, concat_same_type

TomAugspurger commented on pandas/core/indexes/datetimes.py:
```diff
@@ -551,16 +551,12 @@ def snap(self, freq='S'):
+        if level is not None:
+            self._validate_index_level(level)
+
+        # call DatetimeArray method
```
Could you start adding comments like `TODO(DatetimeArray)` here? Once we switch to composition, we should be able to remove the PeriodIndex implementation and use this.

TomAugspurger commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
+    #   * argsort / _values_for_argsort
+    #   * _reduce
+
+    def unique(self):
```
I suppose this closes  https://github.com/pandas-dev/pandas/issues/23586 then? Can you confirm and add a "closes #23586" to the original issue?

TomAugspurger commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
+
+    def copy(self, deep=False):
+        values = self.asi8
+        if deep:
```
@jorisvandenbossche does this look like the correct behavior to you? I was thinking that `.copy()` should break a chain such that in

```python
a = DatetimeArray(...)
b = a.copy()
a[0] = np.nan
```

`b[0]` should not be updated along with `a[0]`. IIUC, with the current implementation `a` and `b` would have the same underlying ndarray (with `deep=False`).

In my head, the `deep` parameter is really only useful when the "scalars" are actually containers (nested data). In this case, `deep=True` would mean recursively copy the values in the "scalar" containers.

jbrockmendel replied:
The `if deep:` branch here was based on jeff's comment [here](https://github.com/pandas-dev/pandas/pull/23415#discussion_r232497536).  I haven't thought about this enough to have a strong opinion about it other than that we shouldn't silently ignore the `deep` kwarg.

jorisvandenbossche replied:
Or simply remove `deep` from the interface? (but that can maybe discussed elsewhere, as it is already part of the interface currenlty)

In any case, I think copy, even with deep=False (the default), should clearly copy the values so that modifying the copy does not change the original one like Tom showed in the example.

TomAugspurger replied:
We have https://github.com/pandas-dev/pandas/issues/22314 for the meaning of `deep`.

TomAugspurger commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        if is_datetime64tz_dtype(original):
```
Can this `if` be replaced with `cls(values, dtype=original.dtype, freq=original.freq)`? `original.dtype` should have the tzinfo in that case.

jbrockmendel replied:
Will do.

jorisvandenbossche replied:
I think this `if` can simply be removed? The values to be factorized are explicitly integers, so `_from_factorized` should always be passed integers.

TomAugspurger commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
```
The problem being that DatetimeArray needs to pass through `tz` info?

For PeriodArray at least (haven't checked TimedeltaArray) you should able to implement `_concat_same_type` just in terms of `.dtype`. It's hashable and can be passed to `PeriodArray.__init__`.

jbrockmendel replied:
Good call

jorisvandenbossche replied:
As opposed to PeriodArray, `freq` is not part of the dtype for DatetimeArray/TimedeltaArray, so I am not sure this check for freq should be done here.

TomAugspurger replied:
IIRC, the PeriodArray constructor allows duplicate `freq` and `dtype`, as long as the match, so

```python
PeriodArray(data, freq='H', dtype=PeriodDtype("H")
```

should be fine.

jbrockmendel replied:
Yah, we added dtype to the PeriodArray constructor specifically so that type(self)(values, freq=self.freq, dtype=self.dtype) would be valid for all three TDA/DTA/PA classes

jorisvandenbossche replied:
Sorry, that's not what I was meaning. The difference is the meaning of `freq` for PeriodArray vs DatetimeArray. For PeriodArray it is part of the dtype and it is defining how the stored integers are interpreted (and thus need to match to just concatenate those integers), but for DatetimeArray it is simply an informative attribute telling you about the regularity of the Array, but not essential to describe it. So I would assume that `_concat_same_type` needs to handle different arrays with different freqs for DatetimeArray.

TomAugspurger commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -139,7 +140,7 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
         return result
 
-    def __new__(cls, values, freq=None):
+    def __new__(cls, values, freq=None, dtype=_TD_DTYPE):
```
(why) was this change necessary?

jbrockmendel replied:
Because in the copy, from_factorized, concat_same_type methods we are passing dtype to this

TomAugspurger replied:
Whoops, I completely missed that the `dtype` argument wasn't present there before. I thought you were changing it from None to `_TD_DTYPE`.

jreback replied:
shouldn't this not still be passed as None and an if check inside? e.g.

```
if dtype is None:
     dtype = _TD_DTYPE
```

as seems passing None should error now

jbrockmendel committed:
Fixup typo, implement more tests

jbrockmendel committed:
whitespace fixup

jbrockmendel committed:
suggested edits

jorisvandenbossche commented on pandas/tests/arrays/test_datetimelike.py:
```diff
@@ -56,7 +56,68 @@ def timedelta_index(request):
+class SharedTests(object):
+    index_cls = None
+
+    def test_take(self):
```
We should not need to add such basic tests I think, as those are covered by the base Extension tests (we should of course test datetime specific aspects).

Is there anything in this test not tested by the base tests?

jbrockmendel replied:
The class-specific tests have tests for invalid fill_values

jorisvandenbossche commented on pandas/tests/arrays/test_datetimelike.py:
```diff
@@ -56,7 +56,68 @@ def timedelta_index(request):
+            arr.take([0, 1], allow_fill=True,
+                     fill_value=pd.Timestamp.now().time)
+
+    def test_concat_same_type(self):
```
As I think also commented on one of the previous PRs that started doing this, I don't think we should test `_concat_same_type` here directly. It is already tested by the base extension tests and by all the tests that actually use it under the hood.

jbrockmendel replied:
Yah, most of these tests were salvaged from one of those older PRs.  I don't see much downside to having the tests, but am pretty happy to pawn this decision/PR off on Tom

TomAugspurger replied:
We still want the fail causes (different dtypes) here.

At this point, we should be able to simplify `core/dtypes/concat.py::concat_datetimetz` and `DatetimeIndex._concat_same_dtype` right? I'll take a look.

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -196,6 +196,74 @@ def astype(self, dtype, copy=True):
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls(values, dtype=original.dtype, freq=original.freq)
```
Does it make sense to pass `original.freq` here? 
It seems a bit strange as this is creating a new array which does not necessarily have the same order as the original one.

Although in practice, if you *have* a freq, that means you have a regular and unique array to start with, so the factorization is kind of a no-op and the result will still have the same freq? (but might be missing corner cases)

jorisvandenbossche replied:
Thinking of a possible corner case, which is currently actually broken: a sorted factorize of a DatetimeIndex with a negative freq: 

```
In [57]: idx = pd.date_range("2012-01-01", periods=3)

In [58]: idx
Out[58]: DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='D')

In [59]: pd.factorize(idx)
Out[59]: 
(array([0, 1, 2]),
 DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='D'))

In [60]: pd.factorize(idx[::-1])
Out[60]: 
(array([0, 1, 2]),
 DatetimeIndex(['2012-01-03', '2012-01-02', '2012-01-01'], dtype='datetime64[ns]', freq='-1D'))

In [61]: pd.factorize(idx[::-1], sort=True)
Out[61]: 
(array([2, 1, 0]),
 DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='-1D'))
```

jbrockmendel commented:
@TomAugspurger are you still good taking this PR over?  (BTW Travis failures look like unrelated timeouts)

TomAugspurger commented:
Sure will push changes here.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into jbrockmendel-dlike_ea4

TomAugspurger commented:
Ah, I see what you mean now. Yes you're right.

And I suppose we can't even assume that concatenating multiple
DatetimeArrays with the same
freq will end up with a DatetimeIndex with the same freq.

On Tue, Nov 13, 2018 at 2:34 PM Joris Van den Bossche <
notifications@github.com> wrote:

> *@jorisvandenbossche* commented on this pull request.
> ------------------------------
>
> In pandas/core/arrays/datetimelike.py
> <https://github.com/pandas-dev/pandas/pull/23643#discussion_r233211534>:
>
> > +        raise AbstractMethodError(self)
> +
> +    def take(self, indices, allow_fill=False, fill_value=None):
> +        if allow_fill:
> +            fill_value = self._validate_fill_value(fill_value)
> +
> +        new_values = take(self.asi8,
> +                          indices,
> +                          allow_fill=allow_fill,
> +                          fill_value=fill_value)
> +
> +        return type(self)(new_values, dtype=self.dtype)
> +
> +    @classmethod
> +    def _concat_same_type(cls, to_concat):
> +        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
>
> Sorry, that's not what I was meaning. The difference is the meaning of
> freq for PeriodArray vs DatetimeArray. For PeriodArray it is part of the
> dtype and it is defining how the stored integers are interpreted (and thus
> need to match to just concatenate those integers), but for DatetimeArray it
> is simply an informative attribute telling you about the regularity of the
> Array, but not essential to describe it. So I would assume that
> _concat_same_type needs to handle different arrays with different freqs
> for DatetimeArray.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23643#discussion_r233211534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIphnpyu5IlWltCDwJywVRHis1KMFks5uuyzTgaJpZM4YaMQT>
> .
>

jorisvandenbossche commented:
> And I suppose we can't even assume that concatenating multiple DatetimeArrays with the same freq will end up with a DatetimeIndex with the same freq.

Indeed, as that would only be the case if they are nicely consecutive ranges.

TomAugspurger committed:
REF: Simplify concat

_concat._concat_datetimetz -> DatetimeIndex._concat_same_dtype ->
DatetimeArray._concat_same_type

TomAugspurger committed:
copy

TomAugspurger committed:
deduplicate copy

TomAugspurger commented:
Pushed some changes.

1. DatetimelikeArray.copy now copies the underlying data
2. changed the concatenation of DatetimeIndex, hopefully for the better.

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -685,17 +689,21 @@ def _concat_same_dtype(self, to_concat, name):
-            return _concat._concat_datetimetz(to_concat, name)
+            # TODO(DatetimeArray)
+            # - remove the .asi8 here
+            # - remove the _maybe_box_as_values
```
looks like you can remove part of the comment

jreback commented:
just a really minor comment. ok to merge.

TomAugspurger commented:
Going to leave the stale comment for now if that's OK. That whole if / else should be going away soon.

jreback commented:
ok! thanks @jbrockmendel and @TomAugspurger
